### PR TITLE
SceneGraphIterator

### PR DESCRIPTION
**Pull request by codex128**
Small quality-of-life improvement. :)

This pull request adds a class that iterates over the scene graph with the depth-first approach. The advantage to using this over the standard scene traversal is being able to use enhanced for-loops, which allow for functionality lambdas are not capable of. I have found this to be very useful for processing scenes at runtime.

Basic usage:
```java
Spatial main = ...
for (Spatial spatial : new SceneGraphIterator(main)) {
    // ...
}
```
The iterator has a couple useful methods:
```java
// makes the iterator not iterate over the children of the current spatial
iterator.ignoreChildren();

// gets the current depth (distance from the current spatial to the main spatial)
iterator.getDepth();
```
Example: count geometries
```java
int total = 0;
Spatial gameScene = rootNode.getChild("game-scene");
for (Spatial spatial : new SceneGraphIterator(gameScene)) {
    if (spatial instanceof Geometry) {
        total++;
    }
}
```
Example: selective material replacement
```java
SceneGraphIterator iterator = new SceneGraphIterator(rootNode);
Material replace = assetManager.loadMaterial("Material/MyMaterial.j3m");
for (Spatial spatial : iterator) {
    // any geometry under "house1" should not use the replacement material
    if ("house1".equals(spatial.getName())) {
        iterator.ignoreChildren();
        continue;
    }
    if (spatial instanceof Geometry) {
        ((Geometry)spatial).setMaterial(replace);
    }
}
```

### CODE REVIEW

File: jme3-examples/src/main/java/jme3test/scene/TestSceneIteration.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    }
+    
+    private Geometry createGeometry(String name) {
+        Geometry g = new Geometry(name, new Box(1, 1, 1));
+        Material m = new Material(assetManager, "Common/MatDefs/Misc/Unshaded.j3md");
+        m.setColor("Color", ColorRGBA.Blue);
+        g.setMaterial(m);
+        return g;
+    }
+    private String constructTabs(int n) {
```
Comment (stephengold) [inline jme3-examples/src/main/java/jme3test/scene/TestSceneIteration.java:73]: A blank line should appear between consecutive methods in a class.

**Review by stephengold — CHANGES_REQUESTED** [review]

**Review by stephengold — APPROVED** [review]

**Comment by stephengold** [comment]
Unless there's further substantive discussion, I plan to integrate this PR in about 48 hours.

**Comment by stephengold** [comment]
GitHub wouldn't let me rebase, so I integrated using "squash and merge".

**Comment by stephengold** [comment]
Thanks for another good contribution to JMonkeyEngine, @codex128 !

**Comment by codex128** [comment]
You're welcome. Glad to do it!

**Pull request merged**